### PR TITLE
feat: 正規表現/JSON整形ツールの複数選択 UI をトグルチップに統一

### PR DESCRIPTION
**Pull request by fumtas1k**
## 概要

「正規表現ビジュアライザ＆ReDoS検出」と「JSON整形・ビューア」で複数選択 UI の語彙がバラバラ（前者＝塗りトグルチップ／後者＝素の checkbox）だったため、共通の**ピル型トグルチップ**に統一しました。あわせて補助テキストの色バグや見出し・余白の不揃いも整えています。

## 変更点

- **共通コンポーネント `ToggleChips` を新規追加**（`src/components/ui/ToggleChips.tsx`）
  - 多選択トグル（`ToggleGroup` の排他選択とは別物）。`<fieldset>`/`<legend>`、各チップは `aria-pressed` ボタン。
  - `token`（文字トークン内包）と `count`（件数バッジ）に対応。
- **正規表現フラグ**: `[g 全マッチ]` のように「文字トークン＋説明」をチップ内に統合し、下の凡例行を撤去。
- **JSON マスク対象**: checkbox → トグルチップ化。検出件数を丸バッジに統合し、別行の「検出: …」を廃止（検出ゼロ時のメッセージのみ残置）。
- **`text-subtle` の色バグ修正**: `@theme` 未定義クラスで本文色になっていた正規表現側の補助テキスト計 12 箇所を `text-muted` に修正。
- **見た目の統一**: グループ見出しを `caption text-muted`（上配置）で両ツール揃え、チップ寸法を 14px で既存トグルと統一。

## アクセシビリティ

- フラグチップは可視テキストと `aria-label` を一致（WCAG 2.5.3 Label in Name）。長文説明は `title` に保持。
- 既存の a11y 回帰ガード（aria-label / title 必須の陽性対照）は名称更新のうえ維持。
- マスク件数は SR 向けに `aria-label`（「メール（検出 N 件）」）へ反映、視覚バッジは `aria-hidden`。

## 検証

- `astro check` 0 errors / Prettier OK / TypeScript OK
- ユニット: Test Files 104 passed / **Tests 1889 passed, 1 skipped**
- E2E: regex-visualizer + json-formatter **33/33 passed**
- build 出力に `.toggle-chip` / `[aria-pressed='true']` / `__token` / `__count` の CSS rule 生成を確認（§7.1 silent regression ガード）
- Playwright MCP で PC(1280x800)・スマホ(390x844) 実機目視（ピル形状・トークン/件数バッジの ON/OFF 視認性・折返し）

## 残作業（マージ後）

- **VRT baseline 更新**: regex-visualizer / json-formatter の見た目が意図的に変わるため、`Update Visual Regression Baseline` workflow の手動 dispatch が必要。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .agents/rules/ui-conventions.md [hunk 1]
```diff
@@ -9,15 +9,16 @@
...
+| `ErrorMessage`        | エラーテキスト表示（`role="alert"` 付き）                                                                                                                                                 |
+| `DownloadButton`      | 統一デザインのダウンロードボタン（アイコン内蔵）                                                                                                                                          |
+| `DownloadButtonGroup` | SVG/PNG ダウンロードボタンペア                                                                                                                                                            |
+| `CopyButton`          | クリップボードコピーボタン                                                                                                                                                                |
+| `ToggleGroup<T>`      | 排他選択トグル（モード切替等）                                                                                                                                                            |
+| `ToggleChips<T>`      | 多選択トグルチップ群。`<fieldset>`/`<legend>` で意味付け、各チップは `aria-pressed` ボタン。`count` prop で件数バッジ表示（マスク検出件数等）、`mono` prop で等幅フォント（フラグ表示等） |
```
Comment (fumtas1k) [inline .agents/rules/ui-conventions.md]: `ToggleChips` の実装には `mono` prop がないため、この説明は将来の利用者を型エラーへ誘導しそうです。フラグの等幅表示は現状 `token` が常に `.toggle-chip__token` で等幅になる設計なので、ここは `token` による等幅トークン表示として書くか、実際に `mono` prop を追加するかのどちらかに揃えたいです。

**Review by fumtas1k — COMMENTED** [review]
多角的に確認しました。

- セキュリティ: 入力データの外部送信追加や CSP を弱める変更はなく、対象 E2E も production CSP 経路で通っています。
- フロントエンドデザイン: 正規表現/JSON の多選択 UI が同じ語彙に寄っており、PC・モバイルともチップの折返し、ON/OFF、件数バッジの視認性は問題なさそうです。VRT fail は regex-visualizer の意図的な見た目差分として扱えますが、マージ前後で baseline 更新は必要です。
- アクセシビリティ: `fieldset`/`legend` と `aria-pressed` の構成は妥当で、フラグの accessible name / title、JSON マスク件数の accessible name もテストで押さえられています。
- ロジック/アーキテクチャ: `ToggleChips` の責務は小さく、既存の `ToggleGroup` と用途が分かれていて妥当です。マスク処理や ReDoS 判定の本体ロジックには不要な変更が入っていません。
- テスト: 変更対象の E2E が陽性対照寄りの観測で更新されており、CI の `test`/`e2e` も pass しています。

ローカル確認:
- `node_modules/.bin/astro check` -> 0 errors / 0 warnings
- `npx playwright test tests/e2e/regex-visualizer.spec.ts tests/e2e/json-formatter.spec.ts` -> 33 passed
- `npm test -- --run tests/meta` -> 18 files / 157 tests passed

指摘は 1 点だけで、実装ブロッカーではなくドキュメント不整合です。

**Comment by fumtas1k** [comment]
ご指摘ありがとうございます。doc を実装の実態に合わせて修正しました（`7b28515`）。

`ToggleChips` に `mono` prop は存在せず、等幅表示は `token`（`.toggle-chip__token`）が常に等幅になる設計で実現しているため、説明を「`token` prop で文字トークンを等幅バッジ表示（フラグ g/i/m 等）」に書き換えました。`mono` prop の新設は YAGNI の観点で見送っています。あわせて本 PR で追加した `legendVisible` prop（見出しを sr-only 化）も同行に追記しました。

**Review by fumtas1k — COMMENTED** [review]
再確認しました。前回指摘した `ToggleChips` の `mono` prop 記述は、実装どおり `token` prop による等幅トークン表示へ修正されており、不整合は解消されています。

あわせて追加された `legendVisible={false}` も確認しました。正規表現フラグの視覚見出しを消しつつ、`legend="正規表現フラグ"` を `sr-only` で a11y ツリーに残す形なので、`fieldset`/`legend` の意味付けを保ったまま余白を戻せていて妥当です。既存の `sr-only` 利用パターンとも整合しています。

追加の指摘はありません。

ローカル確認:
- `node_modules/.bin/astro check` -> 0 errors / 0 warnings
- `npx playwright test tests/e2e/regex-visualizer.spec.ts tests/e2e/json-formatter.spec.ts` -> 33 passed
- `npm run build` -> passed（`BUILD_ID=7b28515` injected）
- `npm test -- --run tests/meta` -> 18 files / 157 tests passed

補足: GitHub checks は確認時点で `no checks reported on the 'feat/unify-toggle-chips-ui' branch` でした。

**Comment by fumtas1k** [comment]
多角的にレビューしました（別セッションの Opus による独立レビュー）。結論として **マージ可能な品質**で、ブロッカーはありません。前回レビューで指摘された doc 不整合（`mono` prop 記述）も `7b28515` で実装に整合済みであることを確認しました。以下、観点別の所見と、任意対応の軽微な指摘 3 点です。

## 総評（観点別）

- **セキュリティ**: 外部送信・CSP 緩和・`dangerouslySetInnerHTML` 追加は無し。`ToggleChips` は受け取った値をテキストノードとして描画するのみで XSS 経路なし。E2E が `withProductionCsp` 経路で通っており本番 CSP 下の挙動も担保されている。問題なし。
- **アーキテクチャ / ロジック**: `ToggleChips`（多選択）と既存 `ToggleGroup`（排他選択）の責務分離が明快。マスク処理・ReDoS 判定の本体ロジックには変更が入っておらず、UI 層のみの差し替えに収まっている。`flag-toggle*` → `toggle-chip*` の CSS 移行も `aria-pressed='true'` セレクタ駆動で `btn-toggle` 流儀に揃っており一貫性がある。
- **フロントエンドデザイン**: 2 ツールの多選択 UI が共通語彙（ピル型チップ＋トークン/件数バッジ）に統一され、狙いどおり。`@layer components` 手書きクラスの hover/focus を擬似クラスで定義しており §7.1 の variant 非対応 silent regression を回避できている。
- **アクセシビリティ**: `fieldset`/`legend` + `aria-pressed` の構成は妥当。フラグの可視ラベル（`short`）と `aria-label` が一致し WCAG 2.5.3（Label in Name）を満たす。`legendVisible={false}` は `sr-only` で legend を a11y ツリーに残しており意味付けを保持。トークン/件数バッジの `aria-hidden` と aria-label への件数反映も整合。
- **テスト**: E2E が `getByText(/検出:/)` → `getByRole('button', { name: /メール（検出/ })` と陽性対照寄りに更新され、「バッジが消えると fail する」検知能力を維持。`astro check` をこちらでも再実行し **0 errors / 0 warnings** を確認。

## 軽微な指摘（任意・ブロッカーではない）

### 1. `ToggleChips` の単体テスト不在（テスト網羅性）
`src/components/ui/` の既存共通コンポーネントは大半が colocated な `.test.tsx` を持っています（`ActionButton` / `InputField` / `DownloadButton` / `FileInputButton` 等）。新規の共有コンポーネントである `ToggleChips` には条件分岐ロジック（`count > 0` でのバッジ表示・`baseLabel \!= null && countVal > 0` の aria-label 合成・`legendVisible` の class 切替・`ariaLabel` 未指定時の label フォールバック）が含まれており、これらは E2E（2 つの consumer 経由）でしか間接的に触れられていません。`ToggleGroup` が未テストの前例はあるものの、ロジックを直接固定する軽量な unit test を追加すると edge case（count=0 で非表示／label が ReactNode のとき aria-label=undefined）を安価に押さえられます。別 issue でも可。

### 2. 検出件数の aria-live アナウンス退行（a11y・軽微）
`JsonMaskControls` で従来は `<p role="status" aria-live="polite">` が検出有無の**両方**で描画され、検出時は「検出: メール 3 …」が SR に読み上げられていました。本 PR では当該 live region が**検出ゼロ時のみ**描画に変わり、検出件数はチップの `aria-label`（`メール（検出 N 件）`）に移っています。aria-label の動的変化は live region と違い自動読み上げされないため、「マスクで N 件検出された」という動的変化が SR ユーザーに通知されなくなります。チップへフォーカス移動すれば件数は読めるので実害は小さいですが、もし動的通知を維持したい場合は件数サマリを `aria-live` 要素に残す選択肢があります（設計判断として許容なら現状で問題なし）。

### 3. stale コメント（軽微）
`RegexVisualizer.tsx:17` のコメント `short: 画面下の凡例用（コンパクト）` は、本 PR で凡例 `<ul>` を撤去し `short` をチップの可視ラベルへ転用したため実態とズレています。`short: チップの可視ラベル用` 等へ更新すると正確です。

---

検証ログ（ローカル）:
- `node_modules/.bin/astro check` -> 0 errors / 0 warnings / 0 hints
- `text-subtle` 修正の裏取り: `--color-subtle` トークン・`.text-subtle` クラスは未定義で、`text-subtle` は無効クラス（色未適用＝本文色フォールバック）。本 PR の `text-muted`（`--color-muted` 経由）への置換は妥当で、`src/` 内に `text-subtle` の残存なしを確認。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by fumtas1k** [comment]
独立レビューありがとうございます。任意指摘 3 点とも本 PR で対応しました。

- **2. 検出件数の aria-live 退行**（`fc0d5a3`）: a11y 退行と判断し修正しました。視覚はチップ内バッジを維持したまま、常設の `sr-only` live region（`role="status" aria-live="polite"`）で件数サマリ（「メール1件、…を検出しました。」）を読み上げます。検出ゼロ時の可視メッセージは `aria-hidden` 化して二重読み上げを防止。
- **1. `ToggleChips` の unit test 不在**（`ea31819`）: `src/components/ui/__tests__/ToggleChips.test.tsx`（15 ケース）を追加。件数バッジ表示・aria-label 合成・`legendVisible` のクラス切替・`ariaLabel` フォールバック・`token` 描画・`aria-pressed`/クリックを直接固定。あわせて #2 の退行ガードとして、件数アナウンスが消えると fail する E2E 陽性対照も追加しました。
- **3. stale コメント**（`efffbe1`）: `RegexVisualizer.tsx` の `short: 画面下の凡例用` を `short: チップの可視ラベル用` に更新。

検証:
- `astro check` -> 0 errors / 0 warnings / 0 hints
- ユニット: Test Files 105 passed / Tests 1906 passed, 1 skipped（ToggleChips 15 passed）
- E2E: json-formatter 21 passed（aria-live 退行ガード含む）/ regex-visualizer pass

VRT baseline（`f694080`）は CI Linux runner で再生成済みです。

**Comment by fumtas1k** [comment]
再レビューしました（前回独立レビューの 3 指摘への対応確認）。**3 件すべて適切に解消**されており、追加の指摘はありません。マージ可能な状態と判断します。

## 指摘対応の確認

### ① `ToggleChips` の単体テスト不在 → 解消（`ea31819`）
`src/components/ui/__tests__/ToggleChips.test.tsx`（15 ケース）を追加。指摘した分岐がすべて固定されています:
- `count > 0` でバッジ表示＋aria-label「ラベル（検出 N 件）」合成 / `count = 0`・未指定でバッジ非表示
- `ariaLabel` 明示時の優先・未指定かつ label 文字列時のフォールバック・**label が ReactNode のとき aria-label=undefined（要素は描画）** の edge case まで網羅
- `legendVisible` の `caption text-muted` ↔ `sr-only` 切替 / `token` 有無 / `aria-pressed` 反映 / クリックで `onToggle(value)`

ローカルで `npx vitest run` 実行 → **15 passed** を確認。

### ② 検出件数の aria-live アナウンス退行 → 解消（`fc0d5a3`）
常設の `sr-only` `role="status" aria-live="polite"` live region を設け、検出時は「メール3件、…を検出しました。」を読み上げ、視覚はチップ内バッジを維持。ゼロ時の可視メッセージは `aria-hidden` 化して二重通知を防いでおり、対応として的確です。E2E 退行ガード（`aria-live` を消すと fail）も併設されており、`test-gates` の陽性対照要件を満たしています。

### ③ stale コメント → 解消（`efffbe1`）
`RegexVisualizer.tsx:14` を `short: チップの可視ラベル用（コンパクト）` に修正。実態と一致。

## 検証（ローカル・`efffbe1`）
- `node_modules/.bin/astro check` → **0 errors / 0 warnings / 0 hints**
- `npx vitest run ToggleChips.test.tsx` → **15 passed**
- `npx playwright test tests/e2e/json-formatter.spec.ts`（aria-live 退行ガード含む）→ **21 passed**

LGTM。マージ後の VRT baseline 更新（PR 本文記載）のみ残作業として認識しています。

🤖 Generated with [Claude Code](https://claude.com/claude-code)
